HPXMLtoOpenStudio/resources/hvac_sizing.rb

```diff
@@ -2684,325 +2668,99 @@ def self.gshp_hxbore_ft_per_ton(weather, hvac_cooling_ap, bore_spacing, pipe_r_v
-        end
-      elsif num_bore_holes == 10
-        if spacing_to_depth_ratio <= 0.02
-          gfnc_coeff = 2.685, 3.08, 3.553, 4.453, 5.552, 7.282, 9.472, 11.405, 14.111, 15.737, 16.888, 18.476, 19.562, 21.966, 24.021, 24.579, 25.086, 25.583, 25.973, 26.311, 26.606, 26.855, 27.043, 27.26, 27.409, 27.482, 27.605
-        elsif spacing_to_depth_ratio <= 0.03
-          gfnc_coeff = 2.679, 3.027, 3.355, 3.871, 4.545, 5.706, 7.332, 8.863, 11.218, 12.688, 13.749, 15.242, 16.284, 18.618, 20.613, 21.161, 21.652, 22.138, 22.521, 22.847, 23.133, 23.376, 23.56, 23.771, 23.915, 23.985, 24.1
-        elsif spacing_to_depth_ratio <= 0.05
-          gfnc_coeff = 2.679, 3.023, 3.319, 3.676, 4.036, 4.645, 5.603, 6.543, 8.285, 9.449, 10.332, 11.623, 12.553, 14.682, 16.613, 17.143, 17.624, 18.094, 18.462, 18.78, 19.057, 19.293, 19.47, 19.673, 19.811, 19.879, 19.989
-        elsif spacing_to_depth_ratio <= 0.1
-          gfnc_coeff = 2.679, 3.023, 3.318, 3.664, 3.962, 4.315, 4.722, 5.045, 5.885, 6.543, 7.086, 7.954, 8.621, 10.291, 11.988, 12.473, 12.931, 13.371, 13.727, 14.03, 14.299, 14.527, 14.698, 14.894, 15.027, 15.092, 15.199
-        else
-          gfnc_coeff = 2.679, 3.023, 3.318, 3.664, 3.961, 4.307, 4.653, 4.842, 5.329, 5.725, 6.069, 6.651, 7.126, 8.478, 9.863, 10.298, 10.704, 11.117, 11.463, 11.755, 12.016, 12.239, 12.407, 12.602, 12.735, 12.8, 12.906
+      x = actuals['b_over_h']
+      x0 = pt1['b_over_h']
+      x1 = pt2['b_over_h']
+      g_functions = gs[0].zip(gs[1]).map { |v| MathTools.interp2(x, x0, x1, v[0], v[1]) }
+
+      # linear interpolation on rb/h for correction factor
+      x = actuals['b_over_h']
+      x0 = pt1['b_over_h']
+      x1 = pt2['b_over_h']
+      f0 = pt1['rb_over_h']
+      f1 = pt2['rb_over_h']
+      actuals['rb_over_h'] = MathTools.interp2(x, x0, x1, f0, f1)
+      rb = actuals['rb_over_h'] * actuals['h']
+      rb_actual_over_rb = actuals['rb'] / rb
+      correction_factor = Math.log(rb_actual_over_rb)
+      g_functions = g_functions.map { |v| v - correction_factor }
+
+      fail "Found different logtimes for '#{bore_config}' with H=#{h1} and H=#{h2}." if logtimes[0] != logtimes[1]
+
+      return logtimes[0], g_functions
+    end
+
+    fail "Could not find gfnc_coeff from '#{g_functions_filename}'."
+  end
+
+  def self.get_g_functions(g_functions_json, bore_config, num_bore_holes, b_h_rb)
+    g_functions_json.each do |_key_1, values_1|
+      if [HPXML::GeothermalLoopBorefieldConfigurationRectangle,
+          HPXML::GeothermalLoopBorefieldConfigurationL].include?(bore_config)
+        bore_locations = values_1[:bore_locations]
+        next if bore_locations.size != num_bore_holes
+
+        logtime = values_1[:logtime].map { |v| Float(v) }
+        g = values_1[:g][b_h_rb.to_sym].map { |v| Float(v) }
+
+        return logtime, g
+      elsif [HPXML::GeothermalLoopBorefieldConfigurationOpenRectangle,
```

Comment:
do we need to do "elseif" instead of "else"? If we do we should probably fail and throw a useful error if someone enters something not in either list

Reply:
The schematron file protects against any invalid borefield configurations. I updated the `test_schema_schematron_error_messages` test to check that it's working properly.